app/src/main/java/com/ncnf/database/DatabaseService.java

```diff
@@ -1,125 +1,245 @@
             }
         });
+
+        return futureResponse;
     }
 
     @Override
-    public CompletableFuture<DatabaseResponse> getData(String path) {
-        CompletableFuture<DatabaseResponse> futureResponse = new CompletableFuture<>();
+    public <T> CompletableFuture<Boolean> setDocument(String documentPath, T object) {
+        CompletableFuture<Boolean> futureResponse = new CompletableFuture<>();
+
+        Map<String, Object> data = registry.get(object.getClass()).toMap(object);
 
-        this.db.document(path).get().addOnCompleteListener(task -> onTaskComplete(task, futureResponse));
+        this.db.document(documentPath).set(data).addOnCompleteListener(task -> {
+            if(task.isSuccessful()){
+                futureResponse.complete(true);
+            } else {
+                futureResponse.completeExceptionally(task.getException());
+            }
+        });
 
         return futureResponse;
     }
 
-    public CompletableFuture<DatabaseResponse> delete(String path){
-        CompletableFuture<DatabaseResponse> futureResponse = new CompletableFuture<>();
 
-        this.db.document(path).delete().addOnCompleteListener(task -> onTaskComplete(task, futureResponse));
+    @Override
+    public <T> CompletableFuture<List<T>> getCollection(String collectionPath, Class<T> collectionType) {
+        CompletableFuture<List<T>> futureResponse = new CompletableFuture<>();
+
+        Task<QuerySnapshot> query = this.db.collection(collectionPath).get();
+        query.addOnCompleteListener(task -> {
+            if(task.isSuccessful()){
+                List<T> result = new ArrayList<>();
+                for(DocumentSnapshot document :  task.getResult().getDocuments()){
+                    result.add((T) registry.get(collectionType).toObject(document.getId(), document.getData()));
+                }
+                futureResponse.complete(result);
+            } else {
+                futureResponse.completeExceptionally(task.getException());
+            }
+        });
 
         return futureResponse;
     }
 
-    public CompletableFuture<DatabaseResponse> updateArrayField(String path, String arrayField, Object value){
-        return this.updateField(path, arrayField, FieldValue.arrayUnion(value));
-    }
+    @Override
+    public <T> CompletableFuture<List<T>> withFieldLike(String collectionPath, String field, String value, Class<T> collectionType) {
+        CompletableFuture<List<T>> futureResponse = new CompletableFuture<>();
+
+        this.db.collection(collectionPath).orderBy(field).startAt(value).endAt(value + "\uf8ff").get().addOnCompleteListener(task -> {
+            if(task.isSuccessful()){
+                List<T> result = new ArrayList<>();
+                for(DocumentSnapshot document : task.getResult().getDocuments()){
+                    result.add((T) registry.get(collectionType).toObject(document.getId(), document.getData()));
```

Comment:
Due to the nature of the result, I would maybe wrap this block in a try/catch to detect potential ClassCastExceptions

Reply:
Also steam+map could help reduce the amount of code and maybe make things clearer

Reply:
Yeah I agree, maybe we'll refactor that later.